ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/CompactionQueryBuilder.java

```diff
@@ -287,16 +302,27 @@ private void buildAddClauseForAlter(StringBuilder query) {
   private void buildSelectClauseForInsert(StringBuilder query) {
     // Need list of columns for major crud, mmmajor partitioned, mmminor
     List<FieldSchema> cols;
-    if (major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
+    if (rebalance || major && crud || major && insertOnly && sourcePartition != null || minor && insertOnly) {
       if (sourceTab == null) {
         return; // avoid NPEs, don't throw an exception but skip this part of the query
       }
       cols = sourceTab.getSd().getCols();
     } else {
       cols = null;
     }
-
-    if (crud) {
+    if (rebalance) {
+      query.append("0, t2.writeId, t2.rowId / CEIL(numRows / ");
```

Comment:
move the query literals into the TxnQueries

Reply:
I have checked TxnQueries and it contains only parameterless whole queries. I think moving these small fragments there would greatly reduce code readability without any real benefit on the other side.